s2/shapeutil_visit_crossing_edge_pairs.go

```diff
@@ -0,0 +1,176 @@
+
+// getShapeEdges returns all edges in the given S2ShapeIndexCell.
+func getShapeEdges(index *ShapeIndex, cell *ShapeIndexCell) ShapeEdgeVector {
+	var shapeEdges ShapeEdgeVector
+	for _, clipped := range cell.shapes {
+		shape := index.Shape(clipped.shapeID)
+		for _, edgeID := range clipped.edges {
+			shapeEdges = append(shapeEdges, ShapeEdge{
+				ID:   ShapeEdgeID{
+					ShapeID: clipped.shapeID,
+					EdgeID: int32(edgeID),
+				},
+				Edge: shape.Edge(edgeID),
+			})
+		}
+	}
+	return shapeEdges
+}
+
+// VisitCrossings finds and processes all crossing edge pairs.
+func visitCrossings(shapeEdges ShapeEdgeVector, crossingType CrossingType, needAdjacent bool, visitor EdgePairVisitor) bool {
+	minCrossingSign := MaybeCross
+	if crossingType == CrossingTypeInterior {
+		minCrossingSign = Cross
+	}
+	for i := 0; i < len(shapeEdges) - 1; i++ {
+		a := shapeEdges[i]
+		j := i + 1
+		// A common situation is that an edge AB is followed by an edge BC.  We
+		// only need to visit such crossings if "needAdjacent" is true (even if
+		// AB and BC belong to different edge chains).
+		if !needAdjacent && a.Edge.V1 == shapeEdges[j].Edge.V0 {
+			j++
+			if j >= len(shapeEdges) {
+				break
+			}
+		}
+		crosser := NewEdgeCrosser(a.Edge.V0, a.Edge.V1)
+		for ; j < len(shapeEdges); j++ {
+			b := shapeEdges[j]
+			if crosser.c != b.Edge.V0 {
+				crosser.RestartAt(b.Edge.V0)
+			}
+			sign := crosser.ChainCrossingSign(b.Edge.V1)
+			// missinglink: enum ordering is reversed compared to C++
+			if sign <= minCrossingSign {
+				if !visitor(a, b, sign == Cross) {
+					return false
+				}
+			}
+		}
+	}
+	return true
+}
+
+// Visits all pairs of crossing edges in the given S2ShapeIndex, terminating
+// early if the given EdgePairVisitor function returns false (in which case
+// VisitCrossings returns false as well).  "type" indicates whether all
+// crossings should be visited, or only interior crossings.
+//
+// If "needAdjacent" is false, then edge pairs of the form (AB, BC) may
+// optionally be ignored (even if the two edges belong to different edge
+// chains).  This option exists for the benefit of FindSelfIntersection(),
+// which does not need such edge pairs (see below).
+func VisitCrossings(index *ShapeIndex, crossingType CrossingType, needAdjacent bool, visitor EdgePairVisitor) bool {
+	// TODO(b/262264880): Use brute force if the total number of edges is small
+	// enough (using a larger threshold if the S2ShapeIndex is not constructed
+	// yet).
+	for it := index.Iterator(); !it.Done(); it.Next() {
+		shapeEdges := getShapeEdges(index, it.cell)
+		if !visitCrossings(shapeEdges, crossingType, needAdjacent, visitor) {
+			return false
+		}
+	}
+	return true
+}
+
+// VisitCrossingEdgePairs finds all crossing edge pairs in an index.
+func VisitCrossingEdgePairs(index *ShapeIndex, crossingType CrossingType, visitor EdgePairVisitor) bool {
+	needAdjacent := crossingType == CrossingTypeAll
+	for it := index.Iterator(); !it.Done(); it.Next() {
```

Comment:
I don't understand why this doesn't call `VisitCrossings` like the C++ does.

https://github.com/google/s2geometry/blob/58de4ea1e2f8a294e0c072c602c22232fd1433ad/src/s2/s2shapeutil_visit_crossing_edge_pairs.cc#L149

Reply:
Two lines below.